### PR TITLE
KAFKA-8855; Collect and Expose Client's Name and Version in the Brokers (KIP-511 Part 1)

### PR DESCRIPTION
**Pull request by dajac**
This PR implements the first part of KIP-511, namely:
- update the ApiVersionsRequest/Response to the auto-generated ones;
- bump the versions of ApiVersionsRequest/Response;
- add ClientSoftwareName and ClientSoftwareVersions in the ApiVersionsRequest;
- add validation for those two fields and handle error in the client;
- implement the proposed fail-back mechanism.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/resources/common/message/ApiVersionsResponse.json [hunk 1]
```diff
@@ -18,14 +18,24 @@
...
+  "validVersions": "0-3",
+  // TODO(dajac) Enable this once KIP-482 is committed.
+  // "flexibleVersions": "3+",
   "fields": [
     { "name": "ErrorCode", "type": "int16", "versions": "0+",
       "about": "The top-level error code." },
     { "name": "ApiKeys", "type": "[]ApiVersionsResponseKey", "versions": "0+",
       "about": "The APIs supported by the broker.", "fields": [
-      { "name": "Index", "type": "int16", "versions": "0+", "mapKey": true,
+      { "name": "ApiKey", "type": "int16", "versions": "0+", "mapKey": true,
```
Comment (dajac) [inline clients/src/main/resources/common/message/ApiVersionsResponse.json:38]: @cmccabe Could we rename this field? `apiKey` is used everywhere at the moment and `index` feels weird. Is it considered as a public API?

File: clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsRequest.java [hunk 2]
```diff
@@ -51,23 +47,28 @@ public Builder(short version) {
...
         }
 
         @Override
         public String toString() {
-            return "(type=ApiVersionsRequest)";
+            return DATA.toString();
         }
     }
 
+    private static final Pattern SOFTWARE_NAME_VERSION_PATTERN = Pattern.compile("(?:[a-zA-Z0-9\\-.])+");
```
Comment (mumrah) [inline clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsRequest.java]: Should we restrict the values for name and version? Maybe we can just test that they are non empty?


nit: the non-capture group isn't really necessary and this regex matches stuff like `"."` and `"---"`.

File: clients/src/main/resources/common/message/ApiVersionsResponse.json (hunk 1 above)
Comment (mumrah) [reply clients/src/main/resources/common/message/ApiVersionsResponse.json:38]: I think anything in the protocols is considered public and would need a KIP.

Comment (dajac) [reply clients/src/main/resources/common/message/ApiVersionsResponse.json:38]: I have noticed that few names have been changed in this PR #7353. The wire format, as described in the documentation, is for sure public. I am not sure about the generator part...

File: clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsRequest.java (hunk 2 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsRequest.java]: We have discussed this in the KIP and the consensus is to have minimal validation (accepted characters) to avoid getting crap from the clients. It is especially important because name and version are used in metrics. 

Good point about the group. Thanks!

Comment (mumrah) [reply clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsRequest.java]: Gotcha, makes sense. Should we maybe check that the first character is alphanumeric?

File: clients/src/main/java/org/apache/kafka/clients/ApiVersion.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ */
+package org.apache.kafka.clients;
+
+import org.apache.kafka.common.message.ApiVersionsResponseData.ApiVersionsResponseKey;
+import org.apache.kafka.common.protocol.ApiKeys;
+
+/**
+ * Represents the min version and max version of an api key.
+ *
+ * NOTE: This class is intended for INTERNAL usage only within Kafka.
```
Comment (stanislavkozlovski) [inline clients/src/main/java/org/apache/kafka/clients/ApiVersion.java:25]: I assume this was based off of our `ApiVersions` class which has the same comment. Maybe some committer will know if there was a reason we didn't create an `kafka.clients.internals` module for such internal classes like we have for other packages

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java [hunk 4]
```diff
@@ -853,18 +854,28 @@ else if (req.isInternalRequest && body instanceof ApiVersionsResponse)
...
-                        apiVersionsResponse.error(), node, req.header.correlationId());
+                        Errors.forCode(apiVersionsResponse.data.errorCode()), node, req.header.correlationId());
                 this.selector.close(node);
                 processDisconnection(responses, node, now, ChannelState.LOCAL_CLOSE);
             } else {
-                nodesNeedingApiVersionsFetch.put(node, new ApiVersionsRequest.Builder((short) 0));
+                // Starting from Apache Kafka 2.4, ApiKeys field is populated with the supported versions of
+                // the ApiVersionsRequest when an UNSUPPORTED_VERSION error is returned.
+                // If not provided, the client fails back to version 0.
+                short failbackVersion = 0;
```
Comment (stanislavkozlovski) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: nit: should this be fallback? I had never heard `failback` before so I googled around a bit - some discussions seem to point it's something different

File: clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java [hunk 5]
```diff
@@ -16,194 +16,115 @@
...
+        // Fallback to version 0 for ApiVersions response. If a client sends an ApiVersionsRequest
+        // using a version higher than that supported by the broker, a version 0 response is sent
+        // to the client indicating UNSUPPORTED_VERSION. When the client receives the response, it
+        // fails back while parsing it into a Struct which means that the version received by this
+        // method is not necessary the real one. It may be version 0 as well.
+        try {
+            return new ApiVersionsResponse(struct, version);
+        } catch (SchemaException e) {
+            if (version != 0)
+                return new ApiVersionsResponse(struct, (short) 0);
```
Comment (stanislavkozlovski) [inline clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java:93]: Do we have a test that exercises this?

File: clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsRequest.java (hunk 2 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsRequest.java]: Sure. I have updated the regex. It checks that the first and last character are alphanumeric.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java (hunk 4 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: I have updated it.

File: clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java (hunk 5 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java:93]: There were not direct unit tests for it but only indirect ones. I have added some for completeness.

File: clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java [hunk 6]
```diff
@@ -73,6 +75,33 @@ public static NodeApiVersions create(Collection<ApiVersion> overrides) {
         return new NodeApiVersions(apiVersions);
     }
 
+    // Used for testing
+    public static NodeApiVersions create(short apiKey, short minVersion, short maxVersion) {
```
Comment (mumrah) [inline clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java:87]: If these are just for unit testing could they be lower visibility? Maybe package-private?

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 7]
```diff
@@ -1616,6 +1616,8 @@ class KafkaApis(val requestChannel: RequestChannel,
       val apiVersionRequest = request.body[ApiVersionsRequest]
       if (apiVersionRequest.hasUnsupportedRequestVersion)
         apiVersionRequest.getErrorResponse(requestThrottleMs, Errors.UNSUPPORTED_VERSION.exception)
+      else if (!apiVersionRequest.isValid)
+        apiVersionRequest.getErrorResponse(requestThrottleMs, Errors.INVALID_REQUEST.exception)
```
Comment (mumrah) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1620]: WDYT about adding a new error code here? If I understand correctly, the client will get this error code when sending a valid ApiVersionsRequest with an invalid client name/version. Something more indicative of the actual problem might be nice.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java [hunk 8]
```diff
@@ -853,18 +854,28 @@ else if (req.isInternalRequest && body instanceof ApiVersionsResponse)
...
-                        apiVersionsResponse.error(), node, req.header.correlationId());
+                        Errors.forCode(apiVersionsResponse.data.errorCode()), node, req.header.correlationId());
                 this.selector.close(node);
                 processDisconnection(responses, node, now, ChannelState.LOCAL_CLOSE);
             } else {
-                nodesNeedingApiVersionsFetch.put(node, new ApiVersionsRequest.Builder((short) 0));
+                // Starting from Apache Kafka 2.4, ApiKeys field is populated with the supported versions of
+                // the ApiVersionsRequest when an UNSUPPORTED_VERSION error is returned.
+                // If not provided, the client falls back to version 0.
+                short fallbackVersion = 0;
```
Comment (mumrah) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: I think it's a little confusing to call this the "fallback" version when really 0 is the fallback version and this variable is holding the max version. Maybe we can rename this to something like `maxApiVersion`?

**Review by mumrah — COMMENTED** [review]
Looks good, a few small nits.

File: clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java (hunk 6 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java:87]: I have checked this and unfortunately it is not possible as it is used by few packages in the clients.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java (hunk 8 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: That makes sense. I have changed it.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 7 above)
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1620]: You have a fair point. Would you go with something like `INVALID_NAME_OR_VERSION` for instance? I have thought about this when I wrote the KIP and went with this error. My reasoning was that we may add new fields in the future (e.g. hostname?) so I wanted to have a generic error which is not tight to the client/version cases. `INVALID_REQUEST` sounds quite appropriate for this.

Furthermore, as the client name and version are set by the developer of a client, it should rarely happen when end users use the client library.

File: clients/src/main/resources/common/message/ApiVersionsResponse.json (hunk 1 above)
Comment (cmccabe) [reply clients/src/main/resources/common/message/ApiVersionsResponse.json:38]: @mumrah: the protocol JSON is not considered a public API.  At some point, that might change, but for now, it's still evolving and can be changed without a KIP.  The protocol itself is public, though.

File: clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java [hunk 9]
```diff
@@ -73,6 +75,23 @@ public static NodeApiVersions create(Collection<ApiVersion> overrides) {
         return new NodeApiVersions(apiVersions);
     }
 
+    // Used for testing
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java]: Let's add some JavaDoc for this function

File: clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java [hunk 10]
```diff
@@ -16,194 +16,115 @@
...
-            new Field(API_VERSIONS_KEY_NAME, new ArrayOf(API_VERSIONS_V0), "API versions supported by the broker."),
-            THROTTLE_TIME_MS);
-
-    /**
-     * The version number is bumped to indicate that on quota violation brokers send out responses before throttling.
-     */
-    private static final Schema API_VERSIONS_RESPONSE_V2 = API_VERSIONS_RESPONSE_V1;
 
     // initialized lazily to avoid circular initialization dependence with ApiKeys
     private static volatile ApiVersionsResponse defaultApiVersionsResponse;
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java]: Please get rid of this volatile.  There is no longer a circular dependency here.

File: clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsRequest.java [hunk 11]
```diff
@@ -78,31 +79,48 @@ public ApiVersionsRequest(short version, Short unsupportedRequestVersion) {
...
-                        version, this.getClass().getSimpleName(), ApiKeys.API_VERSIONS.latestVersion()));
+        ApiVersionsResponseData data = new ApiVersionsResponseData()
+            .setErrorCode(Errors.forException(e).code());
+
+        if (version() >= 1) {
+            data.setThrottleTimeMs(throttleTimeMs);
         }
+
+        // Starting from Apache Kafka 2.4, ApiKeys field is populated with the supported versions of
+        // the ApiVersionsRequest when an UNSUPPORTED_VERSION error is returned.
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsRequest.java]: Add a reference to KIP-511 here

File: clients/src/main/resources/common/message/ApiVersionsResponse.json [hunk 12]
```diff
@@ -18,14 +18,24 @@
...
+  //
   // Starting in version 2, on quota violation, brokers send out responses before throttling.
-  "validVersions": "0-2",
+  //
+  // Version 3 is the first flexible version. Tagged fields are only supported in the body but
+  // not in the header. The length of the header must not change in order to guarantee the
+  // backward compatibility.
+  //
+  // Starting from Apache Kafka 2.4, ApiKeys field is populated with the supported versions of
+  // the ApiVersionsRequest when an UNSUPPORTED_VERSION error is returned.
```
Comment (cmccabe) [inline clients/src/main/resources/common/message/ApiVersionsResponse.json]: Add reference to KIP-511

File: clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java [hunk 13]
```diff
@@ -692,7 +695,44 @@ public void testApiVersionsRequestWithUnsupportedVersion() throws Exception {
...
+        // Send ApiVersionsRequest with a supported version. This should succeed.
+        sendVersionRequestReceiveResponse(node);
+
+        // Test that client can authenticate successfully
+        sendHandshakeRequestReceiveResponse(node, handshakeVersion);
+        authenticateUsingSaslPlainAndCheckConnection(node, handshakeVersion > 0);
+    }
+
+    /**
+     * Tests that invalid ApiVersionRequest V3 is handled by the server correctly and
```
Comment (cmccabe) [inline clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java]: we should not hard-code version 3, but programmatically find the highest supported version and send something higher , so that this code doesn't have to be revisited if there is a new version

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 7 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1620]: This should be pretty rare, so it's probably not worth its own error code.  I do wonder if there should be some kind of log message, since otherwise this would be hard for a broker operator to see.

**Comment by cmccabe** [comment]
Thanks, @dajac .  This looks pretty close to being ready to go.  I left a few comments.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 7 above)
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1620]: I thought about adding a log message as well and I decided to not do it for two reasons:
1) This should be pretty rare and should mainly happen during the development of the client. In this case, the request log could be used; and
2) If a buggy client would be released with an invalid ApiVersionsRequest, I wanted to avoid flooding the logs with something which is not really actionable for a broker operator. What would he do with this information?

File: clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java (hunk 9 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java]: Done.

File: clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java (hunk 10 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java]: Done.

File: clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsRequest.java (hunk 11 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsRequest.java]: Done.

File: clients/src/main/resources/common/message/ApiVersionsResponse.json (hunk 12 above)
Comment (dajac) [reply clients/src/main/resources/common/message/ApiVersionsResponse.json]: Done.

File: clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java (hunk 13 above)
Comment (dajac) [reply clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java]: Done.

**Comment by dajac** [comment]
@cmccabe Thanks for the review. I have addressed all your comments.

File: clients/src/main/java/org/apache/kafka/clients/ApiVersion.java (hunk 3 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/ApiVersion.java:25]: Everything in `org.apache.kafka.clients` is actually internal.  None of those classes are public APIs.  However, the opposite is true for `org.apache.kafka.clients.producer`.  Everything in that namespace is a public API.  So we created `org.apache.kafka.clients.producer.internals` for the non-public producer stuff.

tl;dr: ApiVersion is fine in the `org.apache.kafka.clients` namespace.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 7 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1620]: Those are fair points.  Let's leave it as-is for now.  As you say, the request log will give some hints in this scenario.

**Comment by cmccabe** [comment]
LGTM

**Comment by cmccabe** [comment]
retest this please

**Comment by dajac** [comment]
retest this please

**Comment by dajac** [comment]
Failing tests

JDK 8 and Scala 2.11
* org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest.testAddAndRemoveWorker
* org.apache.kafka.connect.integration.ExampleConnectIntegrationTest.testSourceConnector
* org.apache.kafka.connect.integration.ExampleConnectIntegrationTest.testSinkConnector
* kafka.api.SslAdminClientIntegrationTest.testCreateTopicsResponseMetadataAndConfig

JDK 11 and Scala 2.12
* org.apache.kafka.connect.integration.ExampleConnectIntegrationTest.testSourceConnector
* org.apache.kafka.connect.integration.ExampleConnectIntegrationTest.testSinkConnector

JDK 11 and Scala 2.13
* org.apache.kafka.connect.integration.ExampleConnectIntegrationTest.testSourceConnector
* org.apache.kafka.connect.integration.ExampleConnectIntegrationTest.testSinkConnector

They are unrelated to the changes made in this PR.

**Comment by dajac** [comment]
I ran tests locally and got another failing test:
* ReassignPartitionsClusterTest. shouldPerformMultipleReassignmentOperationsOverVariousTopics

I wanted to confirm that the failing tests in CI are flaky ones.

**Comment by cmccabe** [comment]
Those tests pass for me locally.  Committing

**Comment by hxfghgh** [comment]
hi my friend,i don't know why you have to "add ClientSoftwareName and ClientSoftwareVersions in the ApiVersionsRequest;"。 This does not seem to be a mandatory field to add.
